File: Design-Patterns-Using-JavaScript/ArchitecturalPatterns/MVP.js
// Model
class Model {
    constructor() {
        this.data = "Hello";
    }

    setData(data) {
        console.log("Model: Set data: " + data);
        this.data = data;
    }

    getData() {
        console.log("Model: Get data: " + this.data);
        return this.data;
    }
}

// View
class View {
    getUserInput() {
        console.log("View: getUserInput");
        console.log("View: Enter user input: ");
        const userInput = "Hello, World!";
        console.log(userInput);
        
        return userInput;
    }

    getUserInput2() {
        console.log("View: getUserInput");
        const readline = require('readline').createInterface({
            input: process.stdin,
            output: process.stdout
        });

        readline.question('View : Enter user input: ', userInput => {
            this.controller.handleUserInput(userInput);
            readline.close();
        });
    }

    displayData(data) {
        console.log("View: Display Result: " + data);
    }
}

// Presenter
class Presenter {
    constructor(model, view) {
        this.model = model;
        this.view = view;
    }

    onButtonClick() {
        console.log("Presenter: onButtonClick.");
        const data = this.view.getUserInput();
        this.model.setData(data);
        const newData = this.model.getData();
        this.view.displayData(newData);
    }
}

// Client code
const model = new Model();
const view = new View();
const presenter = new Presenter(model, view);
presenter.onButtonClick();

File: Design-Patterns-Using-Swift/ArchitecturalPatterns/MVP.swift
import Foundation

// Model
class Model {
    private var data: String

    init() {
        self.data = "Hello"
    }

    func setData(_ data: String) {
        print("Model: Set data: \(data)")
        self.data = data
    }

    func getData() -> String {
        print("Model: Get data: \(data)")
        return data
    }
}

// View
class View {
    func getUserInput() -> String {
        print("View: getUserInput")
        print("View: Enter user input: ", terminator: "")
        // let userInput = readLine()
        let userInput =  "Hello, World!"
        print(userInput)
        return userInput
    }

    func displayData(_ data: String) {
        print("View: Display Result: \(data)")
    }
}

// Presenter
class Presenter {
    private var model: Model
    private var view: View

    init(model: Model, view: View) {
        self.model = model
        self.view = view
    }

    func onButtonClick() {
        print("Presenter: onButtonClick.")
        let data = view.getUserInput()
        model.setData(data)
        let newData = model.getData()
        view.displayData(newData)
    }
}

// Client Code
let model = Model()
let view = View()
let presenter = Presenter(model: model, view: view)
presenter.onButtonClick()

/*
Presenter: onButtonClick.
View: getUserInput
View: Enter user input: Hello, World!
Model: Set data: Hello, World!
Model: Get data: Hello, World!
View: Display Result: Hello, World!
*/